vscode/src/completions/context/retrievers/recent-user-actions/recent-view-port.ts

```diff
@@ -0,0 +1,97 @@
+import type { AutocompleteContextSnippet } from '@sourcegraph/cody-shared'
+import { debounce } from 'lodash'
+import * as vscode from 'vscode'
+import type { ContextRetriever } from '../../../types'
+import { RetrieverIdentifier } from '../../utils'
+
+const MAX_RETRIEVED_VIEWPORTS = 5
+
+interface TrackedViewPort {
+    uri: vscode.Uri
+    visibleRange: vscode.Range
+    lastAccessTimestamp: number
+}
+
+export class RecentViewPortRetriever implements vscode.Disposable, ContextRetriever {
+    public identifier = RetrieverIdentifier.RecentViewPortRetriever
+    private disposables: vscode.Disposable[] = []
+    private trackedViewPorts: Map<string, TrackedViewPort> = new Map()
```

Comment:
nit: when using string keys, it can help with clarity to explain what the keys actually are either in the variable name or with a comment
```suggestion
    private viewportsByDocumentUri: Map<string, TrackedViewPort> = new Map()
```

Reply:
done